### ISSUE TITLE
Initializing a `n_transExt` with an `fmpz` broken in Julia 1.3

### ISSUE THREAD

**Issue by fingolfin** [oscar-system/Singular.jl#395]
The following works fine in Julia 1.4, 1.5, 1.6, but is broken in Julia 1.3, on both macOS and Linux/Ubuntu. (See also https://github.com/oscar-system/libsingular-julia/pull/17).
```
julia> using Singular
... current master branch of Singular.jl loads ...

julia> R = FunctionField(QQ, ["a", "b", "c"])[1]
Function Field over Rational Field with transcendence basis n_transExt[a, b, c]

julia> R(3)^80 == R(big(3)^80)
true

julia> R(3)^80 == R(ZZ(3)^80)
true

julia> R(3)^80 == R(Singular.Nemo.ZZ(3)^80)
false
```

Of course we can just decide that we don't care anymore about Julia 1.3... but I really wonder what might cause this?!

**Comment by tthsqe12** [comment]
The relevant source code is here https://github.com/oscar-system/Singular.jl/blob/master/src/libsingular/coeffs.jl#L209

This is a bug in julia/cxxwrap. Here is what is happening with 1.3:
n_InitMPZ_internal is supposed to get a pointer `ptr` to a mpz struct,
but it was ending up with `ptr = first 8 bytes of said mpz struct`.
This means the 4 byte alloc and 4 byte size fields where mashed together an reinterpreted as a byte pointer pointer.
Dereferencing such an unfortunate pointer is a really bad idea.


With julia 1.3, you can see two calls to getindex, which probably is the bug.
```
julia> using Singular

julia> R = FunctionField(QQ, ["a", "b", "c"])[1]
Function Field over Rational Field with transcendence basis n_transExt[a, b, c]

julia> x=Singular.Nemo.ZZ(3)^80
147808829414345923316083210206383297601

julia> code_native(Singular.libSingular.number_ptr, (typeof(x), typeof(R.ptr)), syntax=:intel)
	.text
; ┌ @ coeffs.jl:206 within `CxxPtr'
	push	r15
	push	r14
	push	r12
	push	rbx
	sub	rsp, 152
	mov	r15, rdx
	vxorps	xmm0, xmm0, xmm0
	vmovaps	xmmword ptr [rsp], xmm0
	mov	qword ptr [rsp + 16], 0
	mov	rax, qword ptr fs:[0]
; │┌ @ fmpz.jl:103 within `_fmpz_is_small'
; ││┌ @ Base.jl:20 within `getproperty'
	mov	qword ptr [rsp], 2
	mov	rcx, qword ptr [rax - 15712]
	mov	qword ptr [rsp + 8], rcx
	mov	rcx, rsp
	mov	qword ptr [rax - 15712], rcx
	lea	r14, [rax - 15712]
	mov	rbx, qword ptr [rsi]
; │└└
; │┌ @ promotion.jl:401 within `_fmpz_is_small'
	mov	rax, rbx
	shr	rax, 62
	cmp	rax, 1
; │└
	jne	L232
	movabs	rax, -4611686018427387904
	movabs	rcx, 4611686018427387904
; │ @ coeffs.jl:218 within `CxxPtr'
; │┌ @ gcutils.jl:91 within `macro expansion'
; ││┌ @ fmpz.jl:1986 within `_as_bigint'
; │││┌ @ fmpz.jl:103 within `_fmpz_is_small'
; ││││┌ @ fmpz.jl:119 within `__fmpz_is_small'
; │││││┌ @ operators.jl:193 within `!='
; ││││││┌ @ int.jl:428 within `==' @ promotion.jl:401
	and	rax, qword ptr [rsi]
	cmp	rax, rcx
; │││└└└└
	jne	L328
; ││└
; ││┌ @ coeffs.jl:4 within `n_InitMPZ'
; │││┌ @ CxxWrap.jl:618 within `n_InitMPZ_internal'
	movabs	rbx, offset getindex
	lea	rdi, [rsp + 56]
	mov	esi, 38466496
	mov	edx, 58
	call	rbx             ; first call to getindex
	lea	rdi, [rsp + 32]
	mov	esi, 38466496
	mov	edx, 58             ; second call to getindex ?????
	call	rbx
	mov	rax, qword ptr [rsp + 32]
	test	rax, rax
	je	L400
; ││││┌ @ CxxWrap.jl:572 within `unsafe_convert'
; │││││┌ @ CxxWrap.jl:198 within `CxxPtr'
	mov	rdx, qword ptr [r15]
; ││││└└
	mov	rdi, qword ptr [rsp + 64]
	lea	rsi, [rsp + 128]
	call	rax
	mov	rcx, qword ptr [rsp + 8]
	mov	qword ptr [r14], rcx
; │└└└
	add	rsp, 152
	pop	rbx
	pop	r12
	pop	r14
	pop	r15
	ret
; │ @ coeffs.jl:207 within `CxxPtr'
; │┌ @ CxxWrap.jl:618 within `n_Init'
L232:
	movabs	r12, offset getindex
	lea	rdi, [rsp + 104]
	mov	esi, 38466496
	mov	edx, 54
	call	r12
	lea	rdi, [rsp + 80]
	mov	esi, 38466496
	mov	edx, 54
	call	r12
	mov	rax, qword ptr [rsp + 80]
	test	rax, rax
	je	L422
; ││┌ @ CxxWrap.jl:572 within `unsafe_convert'
; │││┌ @ CxxWrap.jl:198 within `CxxPtr'
	mov	rdx, qword ptr [r15]
; ││└└
	mov	rdi, qword ptr [rsp + 112]
	mov	rsi, rbx
	call	rax
	mov	rcx, qword ptr [rsp + 8]
	mov	qword ptr [r14], rcx
; │└
	add	rsp, 152
	pop	rbx
	pop	r12
	pop	r14
	pop	r15
	ret
; │ @ coeffs.jl:218 within `CxxPtr'
; │┌ @ gcutils.jl:91 within `macro expansion'
; ││┌ @ fmpz.jl:1986 within `_as_bigint'
; │││┌ @ boot.jl:303 within `AssertionError'
L328:
	movabs	rax, offset jl_gc_pool_alloc
	mov	esi, 1400
	mov	edx, 16
	mov	rdi, r14
	call	rax
	movabs	rcx, offset jl_system_image_data
	mov	qword ptr [rax - 8], rcx
	movabs	rcx, 140540651412768
	mov	qword ptr [rax], rcx
	mov	qword ptr [rsp + 16], rax
; │││└
	movabs	rcx, offset jl_throw
	mov	rdi, rax
	call	rcx
; │└└
; │┌ @ CxxWrap.jl:618 within `macro expansion'
L400:
	movabs	rax, offset jl_throw
	movabs	rdi, offset jl_system_image_data
	call	rax
; │└
; │ @ coeffs.jl:207 within `CxxPtr'
; │┌ @ CxxWrap.jl:618 within `n_Init'
L422:
	movabs	rax, offset jl_throw
	movabs	rdi, offset jl_system_image_data
	call	rax
	nop	dword ptr [rax]
; └└
```

With julia 1.6, the code is not perfect but has advanced tremendously from the
stuff that 1.3 output: the 16 byte gmpz struct is read by the
`vmovups xmm0, xmmword ptr [4*rax]`, but then this is put back on the stack
and the stack address is passed along:

```
julia> code_native(Singular.libSingular.number_ptr, (typeof(x), typeof(R.ptr)), syntax=:intel)
	.text
; ┌ @ coeffs.jl:205 within `CxxPtr'
	push	rbp
	mov	rbp, rsp
	sub	rsp, 16
; │ @ coeffs.jl:206 within `CxxPtr'
; │┌ @ fmpz.jl:103 within `_fmpz_is_small'
; ││┌ @ Base.jl:33 within `getproperty'
	mov	rax, qword ptr [rdi]
; ││└
; ││┌ @ fmpz.jl:119 within `__fmpz_is_small'
; │││┌ @ operators.jl:204 within `!='
; ││││┌ @ int.jl:446 within `==' @ promotion.jl:410
	mov	rcx, rax
	shr	rcx, 62
	cmp	ecx, 1
; │└└└└
	jne	L89
; │ @ coeffs.jl:218 within `CxxPtr'
; │┌ @ fmpz.jl:1987 within `_as_bigint'
; ││┌ @ pointer.jl:105 within `unsafe_load' @ pointer.jl:105
	vmovups	xmm0, xmmword ptr [4*rax]
	vmovaps	xmmword ptr [rbp - 16], xmm0
; │└└
; │┌ @ coeffs.jl:4 within `n_InitMPZ'
; ││┌ @ CxxWrap.jl:618 within `n_InitMPZ_internal'
; │││┌ @ array.jl:801 within `getindex'
	cmp	qword ptr [36596744], 57
	jbe	L140
	mov	rax, qword ptr [36596736]
	mov	rcx, qword ptr [rax + 1368]
; │││└
	test	rcx, rcx
	je	L180
; │││┌ @ array.jl:801 within `getindex'
	mov	rdi, qword ptr [rax + 1376]
; │││└
; │││┌ @ CxxWrap.jl:572 within `unsafe_convert'
; ││││┌ @ CxxWrap.jl:198 within `CxxPtr'
	mov	rdx, qword ptr [rsi]
	lea	rsi, [rbp - 16]
; │││└└
	call	rcx
; │└└
	mov	rsp, rbp
	pop	rbp
	ret
; │ @ coeffs.jl:207 within `CxxPtr'
; │┌ @ CxxWrap.jl:618 within `n_Init'
; ││┌ @ array.jl:801 within `getindex'
L89:
	cmp	qword ptr [36596744], 53
	jbe	L202
	mov	rdx, qword ptr [36596736]
	mov	rcx, qword ptr [rdx + 1272]
; ││└
	test	rcx, rcx
	je	L242
; ││┌ @ array.jl:801 within `getindex'
	mov	rdi, qword ptr [rdx + 1280]
; ││└
; ││┌ @ CxxWrap.jl:572 within `unsafe_convert'
; │││┌ @ CxxWrap.jl:198 within `CxxPtr'
	mov	rdx, qword ptr [rsi]
; ││└└
	mov	rsi, rax
	call	rcx
; │└
	mov	rsp, rbp
	pop	rbp
	ret
; │ @ coeffs.jl:218 within `CxxPtr'
; │┌ @ coeffs.jl:4 within `n_InitMPZ'
; ││┌ @ CxxWrap.jl:618 within `n_InitMPZ_internal'
; │││┌ @ array.jl:801 within `getindex'
L140:
	mov	rax, rsp
	lea	rsi, [rax - 16]
	mov	rsp, rsi
	mov	qword ptr [rax - 16], 58
	movabs	rax, offset jl_bounds_error_ints
	mov	edi, 36596736
	mov	edx, 1
	call	rax
; │││└
L180:
	movabs	rax, offset jl_throw
	movabs	rdi, offset jl_system_image_data
	call	rax
; │└└
; │ @ coeffs.jl:207 within `CxxPtr'
; │┌ @ CxxWrap.jl:618 within `n_Init'
; ││┌ @ array.jl:801 within `getindex'
L202:
	mov	rax, rsp
	lea	rsi, [rax - 16]
	mov	rsp, rsi
	mov	qword ptr [rax - 16], 54
	movabs	rax, offset jl_bounds_error_ints
	mov	edi, 36596736
	mov	edx, 1
	call	rax
; ││└
L242:
	movabs	rax, offset jl_throw
	movabs	rdi, offset jl_system_image_data
	call	rax
	nop	dword ptr [rax + rax]
; └└

```

**Comment by tthsqe12** [comment]
After looking at julia 1.3, 1.4, 1.5, and 1.6, I am even more confident that this is a 1.3 bug or our code is just bad. All of 1.4, 1.5, and 1.6 put the target of the unsafe load on the stack and pass along the stack address to libsingular. Version 1.3 also passes along a stack address but the forgets to copy the gmp struct to that stack address.
This is all with the code (note that `n_InitMPZ(b, cf) = n_InitMPZ_internal(pointer_from_objref(b), cf)`):
```
function number_ptr(x::Nemo.fmpz, c::coeffs_ptr)
   if Nemo._fmpz_is_small(x)
      n_Init(x.d, c)
   else
      GC.@preserve x begin
         x1 = Nemo._as_bigint(x)
         y = n_InitMPZ(x1, c)
      end
      return y
   end
end
```

Now the question is whose fault is it that 1.3 forgot to copy the gmp struct. Technically, everything that is made in julia is eventually garbage, so the question is: is `x1` allowed to become garbage before `n_InitMPZ_internal` is called? I see no specification and no details on whether `pointer_from_objref(x1)` holds a reference to `x1` thus preventing it from being collected. Note that this works on 1.3:
```
function number_ptr(x::Nemo.fmpz, c::coeffs_ptr)
   if Nemo._fmpz_is_small(x)
      n_Init(x.d, c)
   else
      GC.@preserve x begin
         x1 = Nemo._as_bigint(x)
         GC.@preserve x1 begin
            y = n_InitMPZ(x1, c)
         end
      end
      return y
   end
end
```

I highly doubt we are going to litter our code with such garbage protection mechanism.

In summary, it seems that is no one's fault - the language is undefined and it is just a miracle that anything works at all :)
help @thofma @rfourquet

**Comment by thofma** [comment]
If we are realling doing `n_InitMPZ(b, cf) = n_InitMPZ_internal(pointer_from_objref(b), cf)`, then this is the culprit. This needs a `GC.@preserve b`, otherwise this is just wrong.

**Comment by tthsqe12** [comment]
Ok, what about the next function in src/libsingular? Do we need gc protection on res? I am inclined to say yes (despite it being in the return). 
```
# get an mpz from a number
function n_GetMPZ(s::number_ptr, r::coeffs_ptr)
   res = BigInt(1)
   resp = pointer_from_objref(res)
   n_GetMPZ_internal(resp, s, r)
   return res
end
```

**Comment by thofma** [comment]
Yes. This is all dodgy.

**Comment by rfourquet** [comment]
> Do we need gc protection on res? I am inclined to say yes (despite it being in the return).

Yes, `res` being the return value is irrelevant for determining whether `GC.@preserve` is required (although of course the current Julia implementation might be such that the code still does what was intended even without `@preseve`).

**Comment by fingolfin** [comment]
Thanks for tracking this down!
